Testerina broken in 1.0.3

**Description:**
Test cases in a ballerina module run fine in 1.0.2 and when upgraded to 1.0.3 the test cases crash with a nullpointer.
Following is found in the ballerina-internal.log
```
[2019-11-14 09:50:21,932] SEVERE {b7a.log.crash} - org.ballerinalang.util.exceptions.BallerinaException: java.lang.NullPointerException 
org.ballerinalang.util.exceptions.BallerinaException: org.ballerinalang.util.exceptions.BallerinaException: java.lang.NullPointerException
	at org.ballerinalang.testerina.core.BTestRunner.formatErrorMessage(BTestRunner.java:811)
	at org.ballerinalang.testerina.core.BTestRunner.lambda$execute$28(BTestRunner.java:656)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at org.ballerinalang.testerina.core.BTestRunner.execute(BTestRunner.java:630)
	at org.ballerinalang.testerina.core.BTestRunner.runTest(BTestRunner.java:160)
	at org.ballerinalang.testerina.util.TesterinaUtils.executeTests(TesterinaUtils.java:132)
	at org.ballerinalang.packerina.task.RunTestsTask.execute(RunTestsTask.java:79)
	at org.ballerinalang.packerina.TaskExecutor.executeTasks(TaskExecutor.java:38)
	at org.ballerinalang.packerina.cmd.BuildCommand.execute(BuildCommand.java:394)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.ballerinalang.tool.Main.main(Main.java:59)
Caused by: org.ballerinalang.util.exceptions.BallerinaException: java.lang.NullPointerException
	at org.ballerinalang.testerina.core.entity.TesterinaFunction.runOnSchedule(TesterinaFunction.java:146)
	at org.ballerinalang.testerina.core.entity.TesterinaFunction.invoke(TesterinaFunction.java:66)
	at org.ballerinalang.testerina.core.entity.TestSuite.start(TestSuite.java:308)
	at org.ballerinalang.testerina.core.BTestRunner.lambda$execute$28(BTestRunner.java:652)
	... 9 more
Caused by: java.lang.NullPointerException
	at org.ballerinalang.jvm.AnnotationUtils.processFPValueAnnotations(AnnotationUtils.java:91)
	at choreo.api.api.choreo_api_0_1_0__testinit_(api:69)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.ballerinalang.testerina.core.entity.TesterinaFunction.lambda$runOnSchedule$0(TesterinaFunction.java:130)
	at org.ballerinalang.jvm.scheduling.SchedulerItem.execute(Scheduler.java:426)
	at org.ballerinalang.jvm.scheduling.Scheduler.run(Scheduler.java:218)
	at org.ballerinalang.jvm.scheduling.Scheduler.runSafely(Scheduler.java:191)
	at java.lang.Thread.run(Thread.java:748)
```
**Steps to reproduce:**
Can be reproduced by using mysql in the code and shutting down mysql server before building the code.

**Affected Versions:**
1.0.3

--- a/cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/RunTestsTask.java
+++ b/cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/RunTestsTask.java
@@ -76,15 +76,14 @@ public class RunTestsTask implements Task {
     }
 
     private HashSet<Path> findDependencyJarPaths(BLangPackage bLangPackage, BuildContext buildContext) {
-
         HashSet<Path> moduleDependencies = buildContext.moduleDependencyPathMap.get(bLangPackage.packageID).
                 platformLibs;
+        if (!bLangPackage.containsTestablePkg()) {
+            return moduleDependencies;
+        }
         // create a new set so that original set is not affected with test dependencies
         HashSet<Path> dependencyJarPaths = new HashSet<>(moduleDependencies);
 
-        if (!bLangPackage.containsTestablePkg()) {
-            return dependencyJarPaths;
-        }
         // add test dependency jars also to the dependency set, if it exists
         BLangTestablePackage testablePackage = bLangPackage.getTestablePkg();
         for (BLangImportPackage importPackage : testablePackage.imports) {
@@ -93,11 +92,11 @@ public class RunTestsTask implements Task {
                 continue;
             }
 
-            // add imported modules dependency jar paths
+            // add imported module's dependent jar paths
             HashSet<Path> testDependencies = buildContext.moduleDependencyPathMap.get(importPkgId).platformLibs;
             dependencyJarPaths.addAll(testDependencies);
 
-            // add imported modules jar path
+            // add imported module's jar path
             Path jarPath = buildContext.getTestJarPathFromTargetCache(importPkgId);
             Path moduleJarPath = buildContext.getJarPathFromTargetCache(importPkgId).getFileName();
             if (Files.exists(jarPath)) {

--- a/cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/RunTestsTask.java
+++ b/cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/RunTestsTask.java
@@ -18,11 +18,14 @@
 
 package org.ballerinalang.packerina.task;
 
+import org.ballerinalang.model.elements.PackageID;
 import org.ballerinalang.packerina.buildcontext.BuildContext;
 import org.ballerinalang.packerina.buildcontext.BuildContextField;
 import org.ballerinalang.testerina.util.TestarinaClassLoader;
 import org.ballerinalang.testerina.util.TesterinaUtils;
+import org.wso2.ballerinalang.compiler.tree.BLangImportPackage;
 import org.wso2.ballerinalang.compiler.tree.BLangPackage;
+import org.wso2.ballerinalang.compiler.tree.BLangTestablePackage;
 
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -61,8 +64,8 @@ public class RunTestsTask implements Task {
             if (Files.notExists(jarPath)) {
                 jarPath = modulejarPath;
             }
-            HashSet<Path> dependencyJarPaths = buildContext.moduleDependencyPathMap
-                    .get(bLangPackage.packageID).platformLibs;
+            // find the set of dependency jar paths for running test for this module
+            HashSet<Path> dependencyJarPaths = findDependencyJarPaths(bLangPackage, buildContext);
             // Create a class loader to run tests.
             TestarinaClassLoader classLoader = new TestarinaClassLoader(jarPath, dependencyJarPaths);
             programFileMap.put(bLangPackage, classLoader);
@@ -70,5 +73,39 @@ public class RunTestsTask implements Task {
         if (programFileMap.size() > 0) {
             TesterinaUtils.executeTests(sourceRootPath, programFileMap, buildContext.out(), buildContext.err());
         }
+    }
+
+    private HashSet<Path> findDependencyJarPaths(BLangPackage bLangPackage, BuildContext buildContext) {
+
+        HashSet<Path> moduleDependencies = buildContext.moduleDependencyPathMap.get(bLangPackage.packageID).
+                platformLibs;
+        // create a new set so that original set is not affected with test dependencies
+        HashSet<Path> dependencyJarPaths = new HashSet<>(moduleDependencies);
+
+        if (!bLangPackage.containsTestablePkg()) {
+            return dependencyJarPaths;
+        }
+        // add test dependency jars also to the dependency set, if it exists
+        BLangTestablePackage testablePackage = bLangPackage.getTestablePkg();
+        for (BLangImportPackage importPackage : testablePackage.imports) {
+            PackageID importPkgId = importPackage.symbol.pkgID;
+            if (!buildContext.moduleDependencyPathMap.containsKey(importPkgId)) {
+                continue;
+            }
+
+            // add imported modules dependency jar paths
+            HashSet<Path> testDependencies = buildContext.moduleDependencyPathMap.get(importPkgId).platformLibs;
+            dependencyJarPaths.addAll(testDependencies);
+
+            // add imported modules jar path
+            Path jarPath = buildContext.getTestJarPathFromTargetCache(importPkgId);
+            Path moduleJarPath = buildContext.getJarPathFromTargetCache(importPkgId).getFileName();
+            if (Files.exists(jarPath)) {
+                dependencyJarPaths.add(jarPath);
+            } else {
+                dependencyJarPaths.add(moduleJarPath);
+            }
+        }
+        return dependencyJarPaths;
     }
 }

--- a/tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/packaging/PathDependencyTestCase.java
+++ b/tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/packaging/PathDependencyTestCase.java
@@ -503,6 +503,22 @@ public class PathDependencyTestCase extends BaseTest {
     }
 
     /**
+     * Build test-dependency project. There is a test time dependency from foo to bar. This should pass as previously,
+     * it was failing to load test time dependency from foo for bar, when test suite of bar was running.
+     *
+     * @throws BallerinaTestException Error when executing the commands.
+     */
+    @Test(description = "Test runtime test dependency from different modules")
+    public void testRuntimeTimeDependencyForExecutingModuleTests() throws BallerinaTestException {
+        Path caseResources = tempTestResources.resolve("test-dependency");
+        String msg = "invoked fooFn";
+        LogLeecher buildLogLeecher = new LogLeecher(msg);
+        balClient.runMain("build", new String[]{"-a"}, envVariables, new String[]{}, new LogLeecher[]{buildLogLeecher},
+                caseResources.toString());
+        buildLogLeecher.waitForText(10000);
+    }
+
+    /**
      * Get environment variables and add ballerina_home as a env variable the tmp directory.
      *
      * @return env directory variable array

--- a/tests/jballerina-integration-test/src/test/resources/packaging/balopath/test-dependency/Ballerina.toml
+++ b/tests/jballerina-integration-test/src/test/resources/packaging/balopath/test-dependency/Ballerina.toml
@@ -1,0 +1,5 @@
+[project]
+org-name= "wso2test"
+version= "0.1.0"
+
+[dependencies]

--- a/tests/jballerina-integration-test/src/test/resources/packaging/balopath/test-dependency/src/bar/main.bal
+++ b/tests/jballerina-integration-test/src/test/resources/packaging/balopath/test-dependency/src/bar/main.bal
@@ -1,0 +1,7 @@
+import ballerina/io;
+
+# Prints `Hello World`.
+
+public function main() {
+    io:println("Hello World!");
+}

--- a/tests/jballerina-integration-test/src/test/resources/packaging/balopath/test-dependency/src/bar/tests/main_test.bal
+++ b/tests/jballerina-integration-test/src/test/resources/packaging/balopath/test-dependency/src/bar/tests/main_test.bal
@@ -1,0 +1,43 @@
+import ballerina/io;
+import ballerina/test;
+import wso2test/foo;
+
+# Before Suite Function
+
+string testStr1 = foo:fooStr1;
+string testStr2 = foo:fooFn();
+
+@test:BeforeSuite
+function beforeSuiteFunc() {
+    io:println("I'm the before suite function!");
+}
+
+# Before test function
+
+function beforeFunc() {
+    io:println("I'm the before function!");
+}
+
+# Test function
+
+@test:Config {
+    before: "beforeFunc",
+    after: "afterFunc"
+}
+function testFunction() {
+    io:println("I'm in test function!");
+    test:assertTrue(true, msg = "Failed!");
+}
+
+# After test function
+
+function afterFunc() {
+    io:println("I'm the after function!");
+}
+
+# After Suite Function
+
+@test:AfterSuite
+function afterSuiteFunc() {
+    io:println("I'm the after suite function!");
+}

--- a/tests/jballerina-integration-test/src/test/resources/packaging/balopath/test-dependency/src/foo/main.bal
+++ b/tests/jballerina-integration-test/src/test/resources/packaging/balopath/test-dependency/src/foo/main.bal
@@ -1,0 +1,14 @@
+import ballerina/io;
+
+# Prints `Hello World`.
+
+public const string fooStr1 = "this is a foo string";
+
+public function main() {
+    io:println("Hello World!");
+}
+
+public function fooFn() returns string {
+    io:println("invoked fooFn");
+    return "fooFn invoked";
+}

--- a/tests/jballerina-integration-test/src/test/resources/packaging/balopath/test-dependency/src/foo/tests/main_test.bal
+++ b/tests/jballerina-integration-test/src/test/resources/packaging/balopath/test-dependency/src/foo/tests/main_test.bal
@@ -1,0 +1,39 @@
+import ballerina/io;
+import ballerina/test;
+
+# Before Suite Function
+
+@test:BeforeSuite
+function beforeSuiteFunc() {
+    io:println("I'm the before suite function!");
+}
+
+# Before test function
+
+function beforeFunc() {
+    io:println("I'm the before function!");
+}
+
+# Test function
+
+@test:Config {
+    before: "beforeFunc",
+    after: "afterFunc"
+}
+function testFunction() {
+    io:println("I'm in test function!");
+    test:assertTrue(true, msg = "Failed!");
+}
+
+# After test function
+
+function afterFunc() {
+    io:println("I'm the after function!");
+}
+
+# After Suite Function
+
+@test:AfterSuite
+function afterSuiteFunc() {
+    io:println("I'm the after suite function!");
+}

